test/extensions/tracers/opencensus/tracer_test.cc

```diff
@@ -161,78 +162,115 @@ TEST(OpenCensusTracerTest, Span) {
-    NiceMock<Tracing::MockConfig> config;
-    Http::TestHeaderMapImpl request_headers{
-        {":path", "/"},
-        {":method", "GET"},
-        {"x-request-id", "foo"},
-        {header, value},
-    };
-    const std::string operation_name{"my_operation_2"};
-    SystemTime start_time;
-    Http::TestHeaderMapImpl injected_headers;
-    {
-      Tracing::SpanPtr span = driver->startSpan(config, request_headers, operation_name, start_time,
-                                                {Tracing::Reason::Sampling, false});
-      span->injectContext(injected_headers);
-      span->finishSpan();
-    }
+namespace {
 
-    // Retrieve SpanData from the OpenCensus trace exporter.
-    std::vector<SpanData> spans = getSpanCatcher()->catchSpans();
-    ASSERT_EQ(1, spans.size());
-    const auto& sd = spans[0];
-    ENVOY_LOG_MISC(debug, "{}", sd.DebugString());
+using testing::PrintToString;
 
-    // Check contents.
-    EXPECT_TRUE(sd.has_remote_parent());
-    EXPECT_EQ("6162636465666768", sd.parent_span_id().ToHex());
-    EXPECT_EQ("404142434445464748494a4b4c4d4e4f", sd.context().trace_id().ToHex());
-    EXPECT_TRUE(sd.context().trace_options().IsSampled())
-        << "parent was sampled, child should be also";
-
-    // Check injectContext.
-    using Envoy::Http::LowerCaseString;
-    {
-      auto val = injected_headers.get(LowerCaseString("traceparent"));
-      ASSERT_NE(nullptr, val);
-      EXPECT_EQ(::opencensus::trace::propagation::ToTraceParentHeader(sd.context()),
-                val->value().getStringView());
-    }
-    {
-      auto val = injected_headers.get(LowerCaseString("grpc-trace-bin"));
-      ASSERT_NE(nullptr, val);
-      std::string expected = ::opencensus::trace::propagation::ToGrpcTraceBinHeader(sd.context());
-      expected = Base64::encode(expected.data(), expected.size(), /*add_padding=*/false);
-      EXPECT_EQ(expected, val->value().getStringView());
-    }
-    {
-      auto val = injected_headers.get(LowerCaseString("x-cloud-trace-context"));
-      ASSERT_NE(nullptr, val);
-      EXPECT_EQ(::opencensus::trace::propagation::ToCloudTraceContextHeader(sd.context()),
-                val->value().getStringView());
-    }
+MATCHER_P2(ContainHeader, header, expected_value,
+           "contains the header " + PrintToString(header) + " with value " +
+               PrintToString(expected_value)) {
+  auto found_value = arg.get(Http::LowerCaseString(header));
+  if (found_value == nullptr) {
+    return false;
+  }
+  return found_value->value().getStringView() == expected_value;
+}
+
+// Given incoming headers, test that trace context propagation works and generates all the expected
+// outgoing headers.
+void TestIncomingHeaders(
+    const std::initializer_list<std::pair<const char*, const char*>>& headers) {
+  registerSpanCatcher();
+  OpenCensusConfig oc_config;
+  NiceMock<LocalInfo::MockLocalInfo> local_info;
+  oc_config.add_incoming_trace_context(OpenCensusConfig::NONE);
```

Comment:
Arguably this could be set in the `TEST` and be a bit narrower.

Reply:
I would prefer to leave this as-is:
* Reduces code duplication in `TEST`s.
* Acts as a smoke-test for incoming headers that aren't found.

As an alternative: how about if I make `TEST` pass the incoming trace context type to `testIncomingHeaders` but still keep this code and all the outgoing contexts (because I want to test translation) in the latter?